library/ssl_tls13_server.c

```diff
@@ -335,6 +335,60 @@ static int ssl_tls13_check_ephemeral_key_exchange( mbedtls_ssl_context *ssl )
     return( 1 );
 }
 
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+/*
+ * Try picking a certificate for this ciphersuite,
+ * return 0 on success and -1 on failure.
+ */
+static int ssl_tls13_pick_cert( mbedtls_ssl_context *ssl )
+{
+    mbedtls_ssl_key_cert *cur, *list;
+
+#if defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)
+    if( ssl->handshake->sni_key_cert != NULL )
+        list = ssl->handshake->sni_key_cert;
+    else
+#endif /* MBEDTLS_SSL_SERVER_NAME_INDICATION */
+        list = ssl->conf->key_cert;
+
+    if( list == NULL )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "server has no certificate" ) );
+        return( -1 );
+    }
+
+    for( cur = list; cur != NULL; cur = cur->next )
+    {
+        MBEDTLS_SSL_DEBUG_CRT( 3, "candidate certificate chain, certificate",
+                               cur->cert );
+
+        /*
+         * This avoids sending the client a cert it'll reject based on
+         * keyUsage or other extensions.
+         */
+        if( mbedtls_x509_crt_check_key_usage( cur->cert, MBEDTLS_X509_KU_DIGITAL_SIGNATURE ) != 0 )
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 3, ( "certificate mismatch: "
+                                      "(extended) key usage extension" ) );
+            continue;
+        }
+
+        break;
```

Comment:
```suggestion

        if( ssl_tls13_get_sig_alg_from_pk( ssl, cur->cert->pk, &algorithm) &&
            mbedtls_ssl_sig_alg_is_received( ssl, algorithm ) )
        {
            ssl->handshake->key_cert = cur;
            MBEDTLS_SSL_DEBUG_CRT( 3, "selected certificate chain, certificate",
                               ssl->handshake->key_cert->cert );
            return( 0 );
        }
        
        
```

Reply:
In last commit: 9850fa8e8d0b0c250114d36d064a1a9821160827, I try to check the certificate key with received signature list, seemd it doesn't work.  Seems the received signature is only used in the protocol.

Reply:
What's the problem ? I think received signature list has been filled when write sni